File: src-tauri/src/commands/cuentas.rs
use sqlx::SqlitePool;
use tauri::State;
use uuid::Uuid;

use crate::api_forward;
use crate::auth;
use crate::models::*;

/// Calcula el nivel de una cuenta a partir de la longitud de su código.
fn calcular_nivel(codigo: &str) -> Result<i32, String> {
    match codigo.len() {
        1 => Ok(1),
        2 => Ok(2),
        3 => Ok(3),
        5 => Ok(4),
        8 => Ok(5),
        _ => Err(format!(
            "Longitud de código inválida ({}). Longitudes válidas: 1 (nivel 1), 2 (nivel 2), 3 (nivel 3), 5 (nivel 4), 8 (nivel 5)",
            codigo.len()
        )),
    }
}

#[tauri::command]
pub async fn listar_cuentas(
    pool: State<'_, SqlitePool>,
    token: String,
) -> Result<Vec<CuentaContableDetalle>, String> {
    let _claims = auth::validate_token(&token)
        .map_err(|e| format!("Token inválido: {}", e))?;

    sqlx::query_as::<_, CuentaContableDetalle>(
        "SELECT c.id, c.id_tipo_cuenta, c.id_cuenta_padre, c.codigo, c.cuenta, c.nivel, c.activo,
                tc.tipo AS tipo_cuenta_nombre,
                cp.codigo AS cuenta_padre_codigo,
                cp.cuenta AS cuenta_padre_nombre
         FROM cuenta_contable c
         INNER JOIN tipo_cuenta tc ON tc.id = c.id_tipo_cuenta
         LEFT JOIN cuenta_contable cp ON cp.id = c.id_cuenta_padre
         ORDER BY c.codigo"
    )
    .fetch_all(pool.inner())
    .await
    .map_err(|e| format!("Error listando cuentas: {}", e))
}

#[tauri::command]
pub async fn crear_cuenta(
    pool: State<'_, SqlitePool>,
    config: State<'_, ApiConfig>,
    auth_token: State<'_, AuthToken>,
    token: String,
    data: CrearCuenta,
) -> Result<CuentaContable, String> {
    let claims = auth::validate_token(&token)
        .map_err(|e| format!("Token inválido: {}", e))?;

    if claims.rol == "lector" {
        return Err("No tiene permisos".to_string());
    }

    let nivel = calcular_nivel(&data.codigo)?;

    // Validar jerarquía si se proporcionó cuenta padre
    if let Some(ref padre_id) = data.id_cuenta_padre {
        let padre = sqlx::query_as::<_, CuentaContable>(
            "SELECT * FROM cuenta_contable WHERE id = ?"
        )
            .bind(padre_id)
            .fetch_optional(pool.inner())
            .await
            .map_err(|e| format!("Error: {}", e))?
            .ok_or_else(|| "Cuenta padre no encontrada".to_string())?;

        if padre.nivel >= nivel {
            return Err(format!(
                "La cuenta padre (nivel {}) debe tener un nivel menor que la cuenta hija (nivel {})",
                padre.nivel, nivel
            ));
        }
    }

    let id = Uuid::new_v4().to_string();

    sqlx::query(
        "INSERT INTO cuenta_contable (id, id_tipo_cuenta, id_cuenta_padre, codigo, cuenta, nivel, sync_status, sync_operation) VALUES (?, ?, ?, ?, ?, ?, 'pending', 'create')"
    )
        .bind(&id)
        .bind(&data.id_tipo_cuenta)
        .bind(&data.id_cuenta_padre)
        .bind(&data.codigo)
        .bind(&data.cuenta)
        .bind(nivel)
        .execute(pool.inner())
        .await
        .map_err(|e| format!("Error creando cuenta: {}", e))?;

    // Forward to REST API — if it succeeds immediately mark as synced
    let synced = api_forward::post(&config, &auth_token, "cuentas", &serde_json::json!({
        "id": id,
        "id_tipo_cuenta": data.id_tipo_cuenta,
        "id_cuenta_padre": data.id_cuenta_padre,
        "codigo": data.codigo,
        "cuenta": data.cuenta,
    })).await;
    if synced {
        sqlx::query("UPDATE cuenta_contable SET sync_status = 'synced', sync_operation = 'none' WHERE id = ?")
            .bind(&id).execute(pool.inner()).await.ok();
    }

    sqlx::query_as::<_, CuentaContable>("SELECT * FROM cuenta_contable WHERE id = ?")
        .bind(&id)
        .fetch_one(pool.inner())
        .await
        .map_err(|e| format!("Error: {}", e))
}

#[tauri::command]
pub async fn editar_cuenta(
    pool: State<'_, SqlitePool>,
    config: State<'_, ApiConfig>,
    auth_token: State<'_, AuthToken>,
    token: String,
    id: String,
    data: EditarCuenta,
) -> Result<CuentaContable, String> {
    let claims = auth::validate_token(&token)
        .map_err(|e| format!("Token inválido: {}", e))?;

    if claims.rol == "lector" {
        return Err("No tiene permisos".to_string());
    }

    let current = sqlx::query_as::<_, CuentaContable>("SELECT * FROM cuenta_contable WHERE id = ?")
        .bind(&id)
        .fetch_optional(pool.inner())
        .await
        .map_err(|e| format!("Error: {}", e))?
        .ok_or_else(|| "Cuenta no encontrada".to_string())?;

    let new_codigo = data.codigo.clone().unwrap_or(current.codigo.clone());
    let new_nivel = calcular_nivel(&new_codigo)?;

    let new_padre = match data.id_cuenta_padre {
        Some(v) => v,
        None => current.id_cuenta_padre.clone(),
    };

    // Validar jerarquía si hay cuenta padre
    if let Some(ref padre_id) = new_padre {
        let padre = sqlx::query_as::<_, CuentaContable>(
            "SELECT * FROM cuenta_contable WHERE id = ?"
        )
            .bind(padre_id)
            .fetch_optional(pool.inner())
            .await
            .map_err(|e| format!("Error: {}", e))?
            .ok_or_else(|| "Cuenta padre no encontrada".to_string())?;

        if padre.nivel >= new_nivel {
            return Err(format!(
                "La cuenta padre (nivel {}) debe tener un nivel menor que la cuenta hija (nivel {})",
                padre.nivel, new_nivel
            ));
        }
    }

    let new_tipo_cuenta = data.id_tipo_cuenta.unwrap_or(current.id_tipo_cuenta.clone());
    let new_cuenta = data.cuenta.unwrap_or(current.cuenta);
    let new_activo = data.activo.unwrap_or(current.activo);

    sqlx::query(
        "UPDATE cuenta_contable SET id_tipo_cuenta = ?, id_cuenta_padre = ?, codigo = ?, cuenta = ?, nivel = ?, activo = ?, sync_status = 'pending', sync_operation = CASE WHEN sync_operation = 'create' THEN 'create' ELSE 'update' END WHERE id = ?"
    )
        .bind(&new_tipo_cuenta)
        .bind(&new_padre)
        .bind(&new_codigo)
        .bind(&new_cuenta)
        .bind(new_nivel)
        .bind(new_activo)
        .bind(&id)
        .execute(pool.inner())
        .await
        .map_err(|e| format!("Error actualizando cuenta: {}", e))?;

    // Forward to REST API — if it succeeds immediately mark as synced
    let synced = api_forward::put(&config, &auth_token, "cuentas", &id, &serde_json::json!({
        "id_tipo_cuenta": new_tipo_cuenta,
        "id_cuenta_padre": new_padre,
        "codigo": new_codigo,
        "cuenta": new_cuenta,
        "activo": new_activo,
    })).await;
    if synced {
        sqlx::query("UPDATE cuenta_contable SET sync_status = 'synced', sync_operation = 'none' WHERE id = ?")
            .bind(&id).execute(pool.inner()).await.ok();
    }

    sqlx::query_as::<_, CuentaContable>("SELECT * FROM cuenta_contable WHERE id = ?")
        .bind(&id)
        .fetch_one(pool.inner())
        .await
        .map_err(|e| format!("Error: {}", e))
}

File: server/src/routes/cuentas.ts
import { Router, Request, Response } from 'express';
import { getPool, RowDataPacket } from '../db';
import { authMiddleware, requireRole } from '../middleware';
import crypto from 'crypto';

const router = Router();
router.use(authMiddleware);

/**
 * Calculate account level from code length.
 * Length 1 → nivel 1, 2 → 2, 3 → 3, 5 → 4, 8 → 5
 */
function calcularNivel(codigo: string): number {
  const map: Record<number, number> = { 1: 1, 2: 2, 3: 3, 5: 4, 8: 5 };
  const nivel = map[codigo.length];
  if (!nivel) {
    throw new Error(
      `Longitud de código inválida (${codigo.length}). Longitudes válidas: 1 (nivel 1), 2 (nivel 2), 3 (nivel 3), 5 (nivel 4), 8 (nivel 5)`
    );
  }
  return nivel;
}

// GET /api/cuentas
router.get('/', async (_req: Request, res: Response) => {
  try {
    const pool = getPool();
    const [rows] = await pool.query<RowDataPacket[]>(
      `SELECT c.id, c.id_tipo_cuenta, c.id_cuenta_padre, c.codigo, c.cuenta, c.nivel, c.activo,
              tc.tipo AS tipo_cuenta_nombre,
              cp.codigo AS cuenta_padre_codigo,
              cp.cuenta AS cuenta_padre_nombre
       FROM cuenta_contable c
       INNER JOIN tipo_cuenta tc ON tc.id = c.id_tipo_cuenta
       LEFT JOIN cuenta_contable cp ON cp.id = c.id_cuenta_padre
       ORDER BY c.codigo`
    );
    res.json(rows);
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

// POST /api/cuentas
router.post('/', requireRole('administrador', 'encargado'), async (req: Request, res: Response) => {
  try {
    const { id_tipo_cuenta, id_cuenta_padre, codigo, cuenta } = req.body;
    const nivel = calcularNivel(codigo);
    const pool = getPool();

    // Validate parent hierarchy
    if (id_cuenta_padre) {
      const [parentRows] = await pool.query<RowDataPacket[]>(
        'SELECT * FROM cuenta_contable WHERE id = ?',
        [id_cuenta_padre]
      );
      if (parentRows.length === 0) {
        res.status(400).json({ error: 'Cuenta padre no encontrada' });
        return;
      }
      if (parentRows[0].nivel >= nivel) {
        res.status(400).json({
          error: `La cuenta padre (nivel ${parentRows[0].nivel}) debe tener un nivel menor que la cuenta hija (nivel ${nivel})`,
        });
        return;
      }
    }

    const id = req.body.id || crypto.randomUUID();
    await pool.query(
      `INSERT INTO cuenta_contable (id, id_tipo_cuenta, id_cuenta_padre, codigo, cuenta, nivel) VALUES (?, ?, ?, ?, ?, ?)
       ON DUPLICATE KEY UPDATE id_tipo_cuenta = VALUES(id_tipo_cuenta), id_cuenta_padre = VALUES(id_cuenta_padre), codigo = VALUES(codigo), cuenta = VALUES(cuenta), nivel = VALUES(nivel)`,
      [id, id_tipo_cuenta, id_cuenta_padre || null, codigo, cuenta, nivel]
    );

    const [rows] = await pool.query<RowDataPacket[]>(
      'SELECT * FROM cuenta_contable WHERE id = ?',
      [id]
    );
    res.status(201).json(rows[0]);
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

// PUT /api/cuentas/:id
router.put('/:id', requireRole('administrador', 'encargado'), async (req: Request, res: Response) => {
  try {
    const { id } = req.params as { id: string };
    const pool = getPool();

    const [existing] = await pool.query<RowDataPacket[]>(
      'SELECT * FROM cuenta_contable WHERE id = ?',
      [id]
    );
    if (existing.length === 0) {
      res.status(404).json({ error: 'Cuenta no encontrada' });
      return;
    }

    const current = existing[0];
    const newCodigo = req.body.codigo ?? current.codigo;
    const newNivel = calcularNivel(newCodigo);
    const newPadre = req.body.id_cuenta_padre !== undefined ? req.body.id_cuenta_padre : current.id_cuenta_padre;

    // Validate parent hierarchy
    if (newPadre) {
      const [parentRows] = await pool.query<RowDataPacket[]>(
        'SELECT * FROM cuenta_contable WHERE id = ?',
        [newPadre]
      );
      if (parentRows.length === 0) {
        res.status(400).json({ error: 'Cuenta padre no encontrada' });
        return;
      }
      if (parentRows[0].nivel >= newNivel) {
        res.status(400).json({
          error: `La cuenta padre (nivel ${parentRows[0].nivel}) debe tener un nivel menor que la cuenta hija (nivel ${newNivel})`,
        });
        return;
      }
    }

    await pool.query(
      'UPDATE cuenta_contable SET id_tipo_cuenta = ?, id_cuenta_padre = ?, codigo = ?, cuenta = ?, nivel = ?, activo = ? WHERE id = ?',
      [
        req.body.id_tipo_cuenta ?? current.id_tipo_cuenta,
        newPadre,
        newCodigo,
        req.body.cuenta ?? current.cuenta,
        newNivel,
        req.body.activo ?? current.activo,
        id,
      ]
    );

    const [rows] = await pool.query<RowDataPacket[]>(
      'SELECT * FROM cuenta_contable WHERE id = ?',
      [id]
    );
    res.json(rows[0]);
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

export default router;
